### DISCUSSION TITLE
Convert Gray Image to RGB format for EfficientAD not working

### DISCUSSION THREAD

**Discussion by Dpat123** [open-edge-platform/anomalib#2811] [Q&A]
i  am trying to convert GrayScale image to RGB for EfficientAD model learning but it is not working 
i have tried following things but none of them working . during model learning it gives error . could you please suggest proper conversion for GrayScale images for use with Anomalib

from anomalib.data.utils.image import read_image, save_image
import numpy as np

                img = read_image(img_path)
                # If the image is 2D, convert it to 3D by stacking the channels
                if img.ndim == 2:
                    img = np.stack((img,) * 3, axis=-1)

                # Unconditionally convert to float32 to ensure compatibility
                img = img.astype(np.float32)
                
                save_image(output_path, img)

**Comment by alfieroddan** [comment]
The `read_image` function already converts GrayScale images on disk to RGB:

```python
 image = Image.open(path).convert("RGB")
```
https://github.com/open-edge-platform/anomalib/blob/3e2906a2953789362a4c11329cb4fbc3d5413761/src/anomalib/data/utils/image.py#L300

Can you show me your full code for running a model on your custom data? Is there an error?

Something like this:

```python
from anomalib.data import Folder
from anomalib.engine import Engine
from anomalib.models import Padim

model = Padim()

datamodule = Folder( #<------- replace with your data here
    name="custom_folder",
    root="./datasets/custom",
    normal_dir="good",
    abnormal_dir="defect"
)

engine = Engine()

engine.fit(
    model=model,
    datamodule=datamodule,
)
```

**Reply by Dpat123** [reply]
Sorry for late reply . At present i have written function that reads file convert it and save it in folder .

`def convert_and_save_images(input_dir, output_dir):
    """
    Convert images from input directory to RGB and save to output directory.
    Includes error handling to skip corrupted or truncated images.
    
    Args:
        input_dir (str): Path to the directory containing input images.
        output_dir (str): Path to the directory where converted images will be saved.
    """
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)
    for filename in os.listdir(input_dir):
        if filename.endswith(('.png', '.jpg', '.jpeg')):
            img_path = Path(input_dir) / filename
            output_path = Path(output_dir) / filename
            try:
                img = read_image(img_path)

                if img is None:
                    print(f"Could not read {filename}, skipping.")
                    continue

                # If the image is 2D, convert it to 3D by stacking the channels
                if img.ndim == 2:
                    img = np.stack((img,) * 3, axis=-1)

                # Unconditionally convert to float32 to ensure compatibility
                img = img.astype(np.float32)
                
                save_image(output_path, img)
                print(f"Converted {filename} and saved to {output_path}")
            except Exception as e:
                print(f"Error processing {filename}: {str(e)}. Skipping this image.")

then in saperate file i try to learn model.

` datamodule = Folder(
        name="top_bottles",
        root=Path("D:/Anomalib/datasets/top_bottles_converted/train"),  
        normal_dir="normal",  # Subfolder containing normal images
        abnormal_dir="anomaly",  # Subfolder containing anomalous images
        train_batch_size=1,  # Set to 1 for debugging, increase for faster training
        eval_batch_size=1,  # Set to 1 for debugging, increase for faster evaluation
        num_workers=0,  # Reduced to 0 to disable multiprocessing and avoid queue timeout issues
    )

    model = EfficientAd(
        teacher_out_channels=384,  # Number of teacher output channels
        model_size= EfficientAdModelSize.S,
        lr=1e-4,
    )

    engine.fit(datamodule=datamodule, model=model)

During learning i get truncated file error.
After reading can we directly use it through folder module instead of writing back ?

**Reply by alfieroddan** [reply]
Yeh give my script a go with your `Folder`, if you run into any issues let me know.

**Reply by Dpat123** [reply]
kindly provide your script to try

**Reply by alfieroddan** [reply]
```python
from anomalib.data import Folder
from anomalib.engine import Engine
from anomalib.models import Padim

model = Padim()

datamodule = Folder( #<------- replace with your data here
    name="custom_folder",
    root="./datasets/custom",
    normal_dir="good",
    abnormal_dir="defect"
)

engine = Engine()

engine.fit(
    model=model,
    datamodule=datamodule,
)
```

**Reply by Dpat123** [reply]
Folder only accept folder path it does not accept any sequence . so how i am going to use read_image and pass that to Folder ?

**Reply by alfieroddan** [reply]
From the `DataModule` documentation: https://anomalib.readthedocs.io/en/latest/markdown/guides/reference/data/datamodules/index.html

You can load your custom data in `Folder` or `Tabular`. Let me show you two ways quickly but next time have a scroll through the docs 😄.

**Folder**
https://anomalib.readthedocs.io/en/latest/markdown/guides/reference/data/datamodules/image/folder.html

The `Folder` data module expects your data to look like this:

```bash
root/
├── normal_dir/
│   ├── image1.png
│   └── image2.png
├── abnormal_dir/
│   ├── image3.png
│   └── image4.png
└── mask_dir/
    ├── mask3.png
    └── mask4.png
```

And then you can load like this:
```python
datamodule = Folder( #<------- replace with your data here
    name="custom_folder",
    root="./datasets/custom",
    normal_dir="good",
    abnormal_dir="defect"
)
```

This is just a standard that `Anomalib` and detests like `MvTec` often use.

**Tabular**
https://anomalib.readthedocs.io/en/latest/markdown/guides/reference/data/datamodules/image/tabular.html

This is where you can use an image sequence. The samples `dict` contains MATCHING paths, label_indexes and split. So the first index of all three lists represent one image's path, label and split.

```python
from anomalib.data import Tabular
from anomalib.data.utils import Split, LabelName

samples = {
    "image_path": ["images/image1.png", "images/image2.png", "images/image3.png", ... ],
    "label_index": [LabelName.NORMAL, LabelName.NORMAL, LabelName.ABNORMAL,  ... ],
    "split": [Split.TRAIN, Split.TRAIN, Split.TEST, ... ],
}
datamodule = Tabular(
    name="custom",
    samples=samples,
    root="./datasets/custom",
)
```

Hope that helps.

**Reply by Dpat123** [reply]
Thanks for the same i will try that.

**Reply by Dpat123** [reply]
I got this working for Gray Images . To setup DataSet as suggest above
` 
    dataset_root = "Anomalib/datasets/top_bottles/train"

    normal_image_paths = sorted([
    str(file_path.relative_to(dataset_root)) for file_path in Path(dataset_root +"/"+ "normal").iterdir()
    ])
    anomalous_image_paths = sorted([
    str(file_path.relative_to(dataset_root)) for file_path in Path(dataset_root +"/"+ "anomaly").iterdir()
    ])

    samples = pd.DataFrame({
    "image_path": normal_image_paths + anomalous_image_paths,    
    "label_index": [0] * len(normal_image_paths) + [1] * len(anomalous_image_paths),
    "split": ["train"] * len(normal_image_paths) + ["test"] * len(anomalous_image_paths),
    })

# Create a Tabular datamodule
    datamodule = Tabular(
    name="TopBottle",
    samples=samples,
    train_batch_size=1,
    eval_batch_size=1,
    root=dataset_root,
    num_workers=1,
    val_split_mode="from_test",  # Create validation set from test set
    val_split_ratio=0.5,  # Use 10% of test set for validation
    )
    datamodule.setup()

# Train images
    train_sample = next(iter(datamodule.train_data))
    val_sample = next(iter(datamodule.val_data))
    print(train_sample.image.shape)
    print(f"Number of validation samples: {len(datamodule.val_data)}")
    print(val_sample.image.shape)
`
